### ISSUE TITLE
Document application plugin's need for its mainClassName property to be set

### ISSUE THREAD

**Issue by rubenaranamorera** [spring-projects/spring-boot#14250]
The documentation states that:
 ```
By default, the executable archive’s main class will be configured automatically by looking for a class with a public static void main(String[]) method in directories on the task’s classpath.
```

But trying a simple example project we are unable to get this autodetected, forcing us to define mainClassName. Is there something I'm missing?

``` 
buildscript {
    repositories {
        mavenCentral()
    }
    dependencies {
        classpath("org.springframework.boot:spring-boot-gradle-plugin:2.0.4.RELEASE")
    }
}

apply plugin: 'java'
apply plugin: 'application'
apply plugin: 'org.springframework.boot'
apply plugin: 'io.spring.dependency-management'

group = 'com.example'
version = '0.0.1-SNAPSHOT'
sourceCompatibility = 1.8

repositories {
    mavenCentral()
}


dependencies {
    compile('org.springframework.boot:spring-boot-starter')
    testCompile('org.springframework.boot:spring-boot-starter-test')
}
```

```
$ ./gradlew build
> Task :startScripts FAILED

FAILURE: Build failed with an exception.

* What went wrong:
A problem was found with the configuration of task ':startScripts'.
> No value has been specified for property 'mainClassName'.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 2s
7 actionable tasks: 3 executed, 4 up-to-date
```

**Comment by philwebb** [comment]
It would be nice if we could add a convention so that the same finder logic is used with the application plugin. If that's not possible we'll need to at least document the restriction.

**Comment by wilkinsona** [comment]
I'm not sure that this is a bug.

The application plugin's own [documentation](https://docs.gradle.org/current/userguide/application_plugin.html) notes that "the only mandatory configuration for the plugin is the specification of the main class (i.e. entry point) of the application" and it's a task from the application plugin that's failing. The documentation [quoted above](
https://github.com/spring-projects/spring-boot/issues/14250#issue-355508735) is specifically referring to the `bootJar` and `bootWar` tasks. We also [already document](https://docs.spring.io/spring-boot/docs/current/gradle-plugin/reference/html/#reacting-to-other-plugins-application) that, when the application plugin is applied, various other tasks (`bootJar`, `bootWar`, and `bootRun`) will then use its `mainClassName` property as a convention.

If we start using the found main class as a convention for the `mainClassName` property we'll either have a cycle between this plugin and the application plugin, or we'll have to change the current behaviour and stop using the `mainClassName` property from the application plugin as a convention for `bootJar`, `bootWar`, and `bootRun`.

In short, I think we should add a note to the documentation [here](https://docs.spring.io/spring-boot/docs/current/gradle-plugin/reference/html/#publishing-your-application-distribution) to draw the user's attention to the fact that the `mainClassName` property is mandatory when using the application plugin.
